File: src/main/java/komachizanj/Komachizan.java
package komachizanj;

import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public class Komachizan {

    private static final Map<Character, Integer> operatorToRank = new HashMap<>();

    static {
        operatorToRank.put('(', 3);
        operatorToRank.put('*', 2);
        operatorToRank.put('/', 2);
        operatorToRank.put('+', 1);
        operatorToRank.put('-', 1);
        operatorToRank.put(')', 0);
    }

    private static Rational calculate(String expression) {

        ArrayDeque<Character> opStack = new ArrayDeque<>();
        ArrayDeque<Rational> varStack = new ArrayDeque<>();
        StringBuilder number = new StringBuilder();
        boolean plusOrMinusSign = false;
        String expr = "(" + expression + ")";

        try {
            for (int i = 0; i < expr.length(); i++) {
                char c = expr.charAt(i);

                if ('0' <= c && c <= '9' || plusOrMinusSign && (c == '+' || c == '-')) {
                    plusOrMinusSign = false;
                    number.append(c);
                    continue;
                }

                plusOrMinusSign = c == '(';

                if (number.length() > 0) {
                    varStack.push(new Rational(Long.parseLong(number.toString())));
                    number.setLength(0);
                }

                while (!opStack.isEmpty()
                        && opStack.peek() != '('
                        && operatorToRank.get(opStack.peek()) >= operatorToRank.get(c)) {
                    Character op = opStack.pop();
                    Rational a = varStack.pop();
                    Rational b = varStack.pop();
                    switch (op) {
                        case '*':
                            varStack.push(b.times(a));
                            break;
                        case '/':
                            if (a.getNumerator() == 0) {
                                throw new ArithmeticException("division by zero : " + expression);
                            }
                            varStack.push(b.div(a));
                            break;
                        case '+':
                            varStack.push(b.plus(a));
                            break;
                        case '-':
                            varStack.push(b.minus(a));
                            break;
                        default:
                            throw new IllegalArgumentException("syntax error : " + expression);
                    }
                }

                if (c == ')') {
                    opStack.pop();
                } else {
                    opStack.push(c);
                }
            }

            return varStack.pop();

        } catch (NoSuchElementException e) {
            throw new IllegalArgumentException("syntax error : $expression");
        }
    }

    private static void komachizan() {

        final int target = 2017;
        final Character[] digits = {'1', '2', '3', '4', '5', '6', '7', '8', '9'};
        final String[] operators = {"", "+", "-", "*", "/"};
        final String[] firstOperators = {"", "-"};

        final List<String> first = Arrays.stream(firstOperators).map(op -> op + digits[0]).collect(Collectors.toList());

        final List<String> expressions = Arrays.stream(digits)
                .skip(1)
                .map(num -> Arrays.stream(operators).map(op -> op + num).collect(Collectors.toList()))
                .reduce(first, (acc, list) -> acc.stream().flatMap(a -> list.stream().map(b -> a + b)).collect(Collectors.toList()));

        final Rational answer = new Rational(target);

        final AtomicInteger cnt = new AtomicInteger();

        expressions
                .parallelStream()
                .filter(it -> calculate(it).compareTo(answer) == 0)
                .collect(Collectors.toList())
                .forEach(it -> System.out.println(cnt.incrementAndGet() + " : " + it + " = " + target));
    }

    public static void main(String[] args) {

        final long start = System.currentTimeMillis();
        komachizan();
        final long end = System.currentTimeMillis();

        System.out.printf("%n処理時間 : %.3f [sec]%n", (end - start) / 1000.0);
    }
}

File: src/main/kotlin/komachizankt/Komachizan.kt
package komachizankt

import java.util.*
import java.util.stream.Collectors
import kotlin.system.measureTimeMillis

val operatorToRank: Map<Char, Int> = mapOf(
        '(' to 3,
        '*' to 2,
        '/' to 2,
        '+' to 1,
        '-' to 1,
        ')' to 0
)

/**
 * 数式を計算します。数式は整数と四則演算子(+ - * /)と丸括弧のみ使えます。
 * （例）12+3/(-45)*67/(8+90)
 * @param expression 数式の文字列。
 * @return 計算結果。
 */
fun calculate(expression: String): Rational {

    val opStack = ArrayDeque<Char>()
    val varStack = ArrayDeque<Rational>()
    val number = StringBuilder()
    var plusOrMinusSign = false
    val expr = "($expression)"

    try {
        for (c in expr) {
            if (c in '0'..'9' || plusOrMinusSign && (c == '+' || c == '-')) {
                plusOrMinusSign = false
                number.append(c)
                continue
            }

            plusOrMinusSign = c == '('

            if (number.isNotEmpty()) {
                varStack.push(Rational(number.toString().toLong()))
                number.clear()
            }

            while (opStack.isNotEmpty()
                    && opStack.peek() != '('
                    && requireNotNull(operatorToRank[opStack.peek()]) >= requireNotNull(operatorToRank[c])
            ) {
                val op = opStack.pop()
                val a = varStack.pop()
                val b = varStack.pop()
                when (op) {
                    '*' -> varStack.push(b * a)
                    '/' -> {
                        if (a.numerator == 0L) {
                            throw ArithmeticException("division by zero : $expression")
                        }
                        varStack.push(b / a)
                    }
                    '+' -> varStack.push(b + a)
                    '-' -> varStack.push(b - a)
                    else -> throw IllegalArgumentException("syntax error : $expression")
                }
            }

            if (c == ')') {
                opStack.pop()
            } else {
                opStack.push(c)
            }
        }

        return varStack.pop()

    } catch (e: NoSuchElementException) {
        throw IllegalArgumentException("syntax error : $expression")
    }
}

fun komachizan() {

    val target = 2017
    val digits = "123456789".toCharArray()
    val operators = arrayOf("", "+", "-", "*", "/")
    val firstOperators = arrayOf("", "-")

    // 数字と演算子の組み合わせ
    // [[1, -1], [2, +2, -2, *2, /2], … [9, +9, -9, *9, /9]]]
    val comb: List<List<String>> =
            listOf(firstOperators.map { op -> op + digits.first() }) +
                    digits.drop(1).map { num -> operators.map { op -> "$op$num" } }

    val expressions: List<String> = comb.reduce { acc, list ->
        acc.flatMap { a -> list.map { b -> "$a$b" } }
    }

    val answer = Rational(target)

    expressions
            .parallelStream()
            .filter { calculate(it).compareTo(answer) == 0 }
            .collect(Collectors.toList())
            .forEachIndexed { index, it ->
                println("${index + 1} : $it = $target")
            }
}

fun main() {

    val elapsed = measureTimeMillis {
        komachizan()
    }

    println("%n処理時間 : %.3f [sec]".format(elapsed / 1000.0))
}
